web-automation-agent/src/agent_actions.ts

```diff
@@ -0,0 +1,191 @@
+    }
+
+    if (!elementFoundAndClicked && text) {
+        const link = await page.$x(`//a[contains(text(), '${text}')]`);
+        if (link && link.length) {
+            // @ts-ignore
+            await link[0].click();
+            elementFoundAndClicked = true;
+        }
+    }
+
+    if (!elementFoundAndClicked) {
+        // TODO: Handle this error
+        throw new Error('Element not found');
+    }
+
+    await waitForNavigation(page);
+    await utils.puppeteer.closeCookieModals(page);
+    await tagAllElementsOnPage(page, UNIQUE_ID_ATTRIBUTE);
+    const minHtml = await shrinkHtmlForWebAutomation(page);
+
+    webAgentLog.info(`Clicked on link, current URL: ${page.url()}`, { text, gid, linkFoundByGidSelector, htmlLength: minHtml.length });
+    return maybeShortsTextByTokenLength(`Previous action was: click_element, HTML of current page: ${minHtml}`, 10000);
+}
+
+export async function fillForm(context: AgentBrowserContext, { formData }: { formData: { gid: number, value: string }[]}) {
+    webAgentLog.info('Calling filling form', { formData });
+    const { page } = context;
+    for (const { gid, value } of formData) {
+        const element = await page.$(`[${UNIQUE_ID_ATTRIBUTE}="${gid}"]`);
+        if (element) {
+            await element.type(value.trim());
+        }
+    }
+    webAgentLog.info('Form filled');
+    const submitButton = await page.$('button[type="submit"]');
+    // If the submit button is not presented, submit the form by pressing enter.
+    if (submitButton) {
+        await submitButton.click();
+    } else {
+        await page.keyboard.press('Enter');
+    }
+    await waitForNavigation(page);
+    await utils.puppeteer.closeCookieModals(page);
+    await tagAllElementsOnPage(page, UNIQUE_ID_ATTRIBUTE);
+    const minHtml = await shrinkHtmlForWebAutomation(page);
+    webAgentLog.info(`Form submitted, current URL: ${page.url()}`, { htmlLength: minHtml.length });
+    return maybeShortsTextByTokenLength(`Previous action was: fill_form_and_submit, HTML of current page: ${minHtml}`, 10000);
+}
+
+export async function extractData(context: AgentBrowserContext, { attributesToExtract }: { attributesToExtract: { gid: number, keyName: string }[] }) {
+    webAgentLog.info('Calling extracting data from page', { attributesToExtract });
+    const { page } = context;
+    const extractedData = {};
+    for (const { gid, keyName } of attributesToExtract) {
+        const element = await page.$(`[${UNIQUE_ID_ATTRIBUTE}="${gid}"]`);
+        if (element) {
+            const value = await page.evaluate((el) => el.textContent, element);
+            // @ts-ignore
+            extractedData[keyName] = value && value.trim();
+        }
+    }
+    webAgentLog.info('Data were extracted from page', { extractedData });
+    return `Extracted JSON data from page: ${JSON.stringify(extractedData)}`;
+}
+
+export async function saveOutput(_: AgentBrowserContext, { object }: { object: { key: string, value: string }[] }) {
```

Comment:
We should also teach it to write to dataset, it might be better overall as you can output one or multiple objects

Reply:
yeah, agree with this. tbh it's pretty easy to let it support both if good description is given to the LLM